src/Nethermind/Nethermind.Network.Discovery/Serializers/NeighborsMsgSerializer.cs

```diff
@@ -1,64 +1,71 @@
 ﻿//  Copyright (c) 2021 Demerzel Solutions Limited
 //  This file is part of the Nethermind library.
-// 
+//
 //  The Nethermind library is free software: you can redistribute it and/or modify
 //  it under the terms of the GNU Lesser General Public License as published by
 //  the Free Software Foundation, either version 3 of the License, or
 //  (at your option) any later version.
-// 
+//
 //  The Nethermind library is distributed in the hope that it will be useful,
 //  but WITHOUT ANY WARRANTY; without even the implied warranty of
 //  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 //  GNU Lesser General Public License for more details.
-// 
+//
 //  You should have received a copy of the GNU Lesser General Public License
 //  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
 
 using System.Net;
+using DotNetty.Buffers;
 using Nethermind.Core.Crypto;
 using Nethermind.Crypto;
 using Nethermind.Network.Discovery.Messages;
+using Nethermind.Network.P2P;
 using Nethermind.Serialization.Rlp;
 using Nethermind.Stats.Model;
 
 namespace Nethermind.Network.Discovery.Serializers;
 
-public class NeighborsMsgSerializer : DiscoveryMsgSerializerBase, IMessageSerializer<NeighborsMsg>
+public class NeighborsMsgSerializer : DiscoveryMsgSerializerBase, IZeroMessageSerializer<NeighborsMsg>
 {
     public NeighborsMsgSerializer(IEcdsa ecdsa,
         IPrivateKeyGenerator nodeKey,
         INodeIdResolver nodeIdResolver) : base(ecdsa, nodeKey, nodeIdResolver)
     {
     }
 
-    public byte[] Serialize(NeighborsMsg msg)
+    public void Serialize(IByteBuffer byteBuffer, NeighborsMsg msg)
     {
-        Rlp[]? nodes = null;
+        int length = GetLength(msg, out int contentLength);
+
+        RlpStream stream = new(length);
+        stream.StartSequence(contentLength);
         if (msg.Nodes.Any())
         {
-            nodes = new Rlp[msg.Nodes.Length];
+            int nodeLength = GetNodesLength(msg.Nodes, out int contentLengthNodes);
```

Comment:
you calculate nodes length for the second time.
It would be better to remember the values when done for the first time. So maybe smth like this?
_(int length, int contentLegth, int nodesLength, int nodesContentLength) GetLength(NeighborsMsg msg)_
and remove method GetNodesLength completely?

Reply:
Again, done to just keep the GetLength interface similar to other serializers - its like this for a lot of serializers.
If we decide to not standardize the interface, then I will change this is all the serializers.

Reply:
It's not everywhere like this. I've already used tuples as returns in multiple places to avoid redundant length recalculations.

Reply:
okay, then I will change the methods to use and return tuples

Reply:
Example: _StorageRangesMessageSerializer_

Reply:
we have a interface
```public interface IZeroInnerMessageSerializer<T> : IZeroMessageSerializer<T> where T : MessageBase
    {
        int GetLength(T message, out int contentLength);
    }```

Reply:
which is used to get length of the message in quite some places. so we cannot technically convert to returning tuples

Reply:
we can either have two functions, one that returns tuples and the other that can be used externally to get the message length